python/example_code/cloudwatch-logs/scenarios/large-query/cloudwatch_query.py

```diff
@@ -0,0 +1,193 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+import logging
+import time
+from datetime import datetime
+
+from date_utilities import DateUtilities
+
+
+class DateOutOfBoundsError(Exception):
+    """Exception raised when the date range for a query is out of bounds."""
+
+    pass
+
+
+class CloudWatchQuery:
+    """
+    A class to query AWS CloudWatch logs within a specified date range.
+
+    :ivar client: The Cloudwatch Logs Client.
+    :vartype client: Client
+    :ivar log_groups: Names of the log groups to query.
+    :vartype log_groups: list
+    :ivar date_range: Start and end datetime for the query.
+    :vartype date_range: tuple
+    :ivar limit: Maximum number of log entries to return.
+    :vartype limit: int
+    """
+
+    def __init__(self, client, log_groups, date_range, limit=10000):
+        self.cloudwatch_logs = client
+        self.log_groups = log_groups
+        self.date_range = date_range
+        self.limit = limit
+        self.query_results = []
+        self.query_duration = None
+        self.datetime_format = "%Y-%m-%d %H:%M:%S.%f"
+        self.log_batch = None
+        self.date_utilities = DateUtilities()
+
+    def query_logs(self):
+        """
+        Executes a CloudWatch logs query for a specified date range and calculates the execution time of the query.
+
+        :return: A batch of logs retrieved from the CloudWatch logs query.
+        :rtype: list
+        """
+        start_time = datetime.now()
+
+        start_date, end_date = self.date_utilities.normalize_date_range_format(
+            self.date_range, from_format="unix_timestamp", to_format="datetime"
+        )
+
+        logging.info(
+            f"Original query:"
+            f"\n       START:    {start_date}"
+            f"\n       END:      {end_date}"
+        )
+        logs_batch = self.process_logs((start_date, end_date))
+        end_time = datetime.now()
+        self.query_duration = (end_time - start_time).total_seconds()
+        return logs_batch
+
+    def process_logs(self, date_range):
```

Comment:
This looks like standard pagination instead of a [binary search](https://github.com/awsdocs/aws-doc-sdk-examples/blob/main/workflows/cloudwatch_logs_large_query/SPECIFICATION.md#recursive-queries)

Reply:
Done!